File: android/app/src/main/java/com/example/viewcast/rtsp/RtspServer.kt
package com.example.viewcast.rtsp

import kotlinx.coroutines.*
import java.io.BufferedOutputStream
import java.io.BufferedReader
import java.io.InputStreamReader
import java.net.ServerSocket
import java.net.Socket
import java.nio.ByteBuffer
import android.media.MediaCodec

class RtspServer(private val scope: CoroutineScope, private val port: Int = 8554) {
    private var serverJob: Job? = null
    private val clients = mutableListOf<Client>()

    data class Client(val socket: Socket, val out: BufferedOutputStream)

    @Volatile private var sps: ByteBuffer? = null
    @Volatile private var pps: ByteBuffer? = null

    fun updateSpsPps(spsBuf: ByteBuffer, ppsBuf: ByteBuffer) {
        sps = spsBuf.duplicate()
        pps = ppsBuf.duplicate()
    }

    fun start() {
        serverJob = scope.launch(Dispatchers.IO) {
            val server = ServerSocket(port)
            while (isActive) {
                val socket = server.accept()
                launch { handleClient(socket) }
            }
        }
    }

    fun stop() {
        serverJob?.cancel()
        synchronized(clients) {
            clients.forEach { runCatching { it.socket.close() } }
            clients.clear()
        }
    }

    private suspend fun handleClient(socket: Socket) {
        socket.tcpNoDelay = true
        val reader = BufferedReader(InputStreamReader(socket.getInputStream()))
        val out = BufferedOutputStream(socket.getOutputStream())
        var session = "12345678"

        while (true) {
            val request = readRequest(reader) ?: break
            val (method, url, cseq) = request

            when (method) {
                "OPTIONS" -> {
                    respond(out, cseq, extraHeaders = listOf("Public: OPTIONS, DESCRIBE, SETUP, PLAY"))
                }
                "DESCRIBE" -> {
                    val spsLocal = sps ?: ByteBuffer.allocate(0)
                    val ppsLocal = pps ?: ByteBuffer.allocate(0)
                    val spsB64 = android.util.Base64.encodeToString(byteArrayOf(*spsLocal.copyBytes()), android.util.Base64.NO_WRAP)
                    val ppsB64 = android.util.Base64.encodeToString(byteArrayOf(*ppsLocal.copyBytes()), android.util.Base64.NO_WRAP)
                    val sdp = buildString {
                        append("v=0\r\n")
                        append("o=- 0 0 IN IP4 127.0.0.1\r\n")
                        append("s=AndroidScreen\r\n")
                        append("t=0 0\r\n")
                        append("m=video 0 RTP/AVP 96\r\n")
                        append("c=IN IP4 0.0.0.0\r\n")
                        append("a=rtpmap:96 H264/90000\r\n")
                        append("a=fmtp:96 packetization-mode=1;sprop-parameter-sets=$spsB64,$ppsB64\r\n")
                        append("a=control:trackID=1\r\n")
                    }
                    val headers = listOf(
                        "Content-Base: $url",
                        "Content-Type: application/sdp",
                        "Content-Length: ${sdp.toByteArray().size}"
                    )
                    respond(out, cseq, extraHeaders = headers, body = sdp)
                }
                "SETUP" -> {
                    // Interleaved TCP RTP
                    session = "sess" + System.currentTimeMillis()
                    val headers = listOf(
                        "Transport: RTP/AVP/TCP;unicast;interleaved=0-1",
                        "Session: $session"
                    )
                    respond(out, cseq, extraHeaders = headers)
                    synchronized(clients) { clients.add(Client(socket, out)) }
                }
                "PLAY" -> {
                    val headers = listOf("RTP-Info: url=$url/trackID=1;seq=0")
                    respond(out, cseq, extraHeaders = headers)
                }
                else -> respond(out, cseq)
            }
        }
    }

    private fun respond(out: BufferedOutputStream, cseq: String, extraHeaders: List<String> = emptyList(), body: String? = null) {
        val sb = StringBuilder()
        sb.append("RTSP/1.0 200 OK\r\n")
        sb.append("CSeq: $cseq\r\n")
        extraHeaders.forEach { sb.append("$it\r\n") }
        if (body != null) {
            sb.append("\r\n")
            sb.append(body)
        } else {
            sb.append("\r\n")
        }
        out.write(sb.toString().toByteArray())
        out.flush()
    }

    private fun readRequest(reader: BufferedReader): Triple<String, String, String>? {
        var line = reader.readLine() ?: return null
        if (line.isEmpty()) return null
        val requestLine = line
        var method = ""
        var url = ""
        requestLine.split(" ").let {
            method = it.getOrNull(0) ?: ""
            url = it.getOrNull(1) ?: ""
        }
        var cseq = "1"
        while (true) {
            line = reader.readLine() ?: break
            if (line.isEmpty()) break
            if (line.startsWith("CSeq:", true)) {
                cseq = line.split(":")[1].trim()
            }
        }
        return Triple(method, url, cseq)
    }

    fun sendH264Sample(nal: ByteArray, isKeyFrame: Boolean) {
        val rtpHeader = ByteArray(12)
        // Minimalistic: not a full RTP implementation; for demo only.
        // For TCP interleaving, prepend '$' channel byte and 2-byte length
        val payload = nal
        val interleavedHeader = byteArrayOf('$'.code.toByte(), 0, ((payload.size shr 8) and 0xFF).toByte(), (payload.size and 0xFF).toByte())

        synchronized(clients) {
            val iter = clients.iterator()
            while (iter.hasNext()) {
                val c = iter.next()
                try {
                    c.out.write(interleavedHeader)
                    c.out.write(payload)
                    c.out.flush()
                } catch (e: Exception) {
                    runCatching { c.socket.close() }
                    iter.remove()
                }
            }
        }
    }

    private fun ByteBuffer.copyBytes(): ByteArray {
        val dup = duplicate()
        val arr = ByteArray(dup.remaining())
        dup.get(arr)
        return arr
    }
}

File: ios/ViewCast/RtspServer.swift
import Foundation
import Network

class RtspServer {
    private let listener: NWListener
    private var connections: [NWConnection] = []
    private var sps: Data?
    private var pps: Data?

    init(port: UInt16 = 8554) throws {
        listener = try NWListener(using: .tcp, on: NWEndpoint.Port(rawValue: port)!)
        listener.newConnectionHandler = { [weak self] conn in
            self?.setup(conn)
        }
    }

    func start() { listener.start(queue: .global()) }
    func stop() { listener.cancel(); connections.forEach { $0.cancel() } }

    func updateSpsPps(sps: Data, pps: Data) { self.sps = sps; self.pps = pps }

    private func setup(_ conn: NWConnection) {
        conn.start(queue: .global())
        connections.append(conn)
        receiveRequest(on: conn)
    }

    private func receiveRequest(on conn: NWConnection) {
        conn.receive(minimumIncompleteLength: 1, maximumLength: 8192) { [weak self] data, _, isComplete, err in
            guard let self = self, let d = data, !d.isEmpty else { return }
            let req = String(data: d, encoding: .utf8) ?? ""
            let cseq = req.components(separatedBy: "\r\n").first(where: { $0.hasPrefix("CSeq:") })?.split(separator: ":").last?.trimmingCharacters(in: .whitespaces) ?? "1"
            if req.starts(with: "DESCRIBE") {
                let spsB64 = (self.sps ?? Data()).base64EncodedString()
                let ppsB64 = (self.pps ?? Data()).base64EncodedString()
                let sdp = """
                v=0
                o=- 0 0 IN IP4 127.0.0.1
                s=iOSScreen
                t=0 0
                m=video 0 RTP/AVP 96
                a=rtpmap:96 H264/90000
                a=fmtp:96 packetization-mode=1;sprop-parameter-sets=\(spsB64),\(ppsB64)
                a=control:trackID=1

                """
                let resp = "RTSP/1.0 200 OK\r\nCSeq: \(cseq)\r\nContent-Type: application/sdp\r\nContent-Length: \(sdp.utf8.count)\r\n\r\n\(sdp)"
                conn.send(content: resp.data(using: .utf8), completion: .contentProcessed({ _ in }))
            } else if req.starts(with: "OPTIONS") {
                let resp = "RTSP/1.0 200 OK\r\nCSeq: \(cseq)\r\nPublic: OPTIONS, DESCRIBE, SETUP, PLAY\r\n\r\n"
                conn.send(content: resp.data(using: .utf8), completion: .contentProcessed({ _ in }))
            } else if req.starts(with: "SETUP") {
                let resp = "RTSP/1.0 200 OK\r\nCSeq: \(cseq)\r\nTransport: RTP/AVP/TCP;unicast;interleaved=0-1\r\nSession: 1\r\n\r\n"
                conn.send(content: resp.data(using: .utf8), completion: .contentProcessed({ _ in }))
            } else if req.starts(with: "PLAY") {
                let resp = "RTSP/1.0 200 OK\r\nCSeq: \(cseq)\r\nRTP-Info: url=rtsp://localhost/trackID=1;seq=0\r\n\r\n"
                conn.send(content: resp.data(using: .utf8), completion: .contentProcessed({ _ in }))
            } else if req.starts(with: "PAUSE") {
                let resp = "RTSP/1.0 200 OK\r\nCSeq: \(cseq)\r\n\r\n"
                conn.send(content: resp.data(using: .utf8), completion: .contentProcessed({ _ in }))
            } else if req.starts(with: "TEARDOWN") {
                let resp = "RTSP/1.0 200 OK\r\nCSeq: \(cseq)\r\n\r\n"
                conn.send(content: resp.data(using: .utf8), completion: .contentProcessed({ _ in }))
                // Close connection
                conn.cancel()
            }
            if !isComplete { self.receiveRequest(on: conn) }
        }
    }

    func sendNal(_ nal: Data) {
        let header: [UInt8] = [UInt8(ascii: "$"), 0, UInt8((nal.count >> 8) & 0xFF), UInt8(nal.count & 0xFF)]
        let pkt = Data(header) + nal
        for c in connections {
            c.send(content: pkt, completion: .contentProcessed({ _ in }))
        }
    }
}
